### ISSUE TITLE
Segmentation fault in SeriesCapture + ProteanChess

### ISSUE THREAD

**Issue by thomas-maeder** [thomas-maeder/popeye#433]
begin
author Bruno Stucker
pieces white bh8 pc3 black ka1 bd4 pa2b1
stip h#1
condition protean seriescapture
option nowk
twin move c3 b2
end

gives

a) 

  1.Bd4*h8-d4 c3*d4=B #

b) wPc3-->b2  
Segmentation fault

**Comment by thomas-maeder** [comment]
Another example:

begin
author Bruno Stucker
pieces white rc4 bf5 pd3e4 black ka1 rd2 bf4 pa2
stip h#1
condition protean seriescapture
option nowk
twin add white pc2
end

a) 

  1.Rd2*d3=PP*e4*f5=B-b1 Rc4*f4=B-e5 #
  add_to_move_generation_stack:        1050
                     play_move:         992
 is_white_king_square_attacked:           0
 is_black_king_square_attacked:        1022

b) +wPc2  

  1.Rd2*c2=PP*d3*e4*f5=B-b1 Rc4*f4=B-e5 # Segmentation fault

**Comment by JoshuaGreen** [comment]
At least one problem is that in lines 1479-1480  [`/optimisations/hash.c`](https://github.com/thomas-maeder/popeye/blob/develop/optimisations/hash.c) the range of `i` (from `en_passant_top[nbply-1]+1` to `en_passant_top[nbply]`, inclusive) is taking us outside the `en_passant_multistep_over` array.

**Comment by JoshuaGreen** [comment]
Further information &mdash; we're reaching line 1479 with `en_passant_top[nbply] == UINT_MAX`.

**Comment by JoshuaGreen** [comment]
In fact, I suspect we should be able to [`assert`](https://en.wikipedia.org/wiki/Assert.h) that `en_passant_top[nbply] > en_passant_top[nbply-1]` before decrementing `en_passant_top[nbply]`.

**Comment by JoshuaGreen** [comment]
It does seem that replacing line 190 of [`pieces/walks/pawns/en_passant.c`](https://github.com/thomas-maeder/popeye/blob/develop/pieces/walks/pawns/en_passant.c) with
```c
if (en_passant_was_multistep_played(nbply))
  --en_passant_top[nbply];
```
eliminates the segfaults without breaking any of the non-long regression tests.&nbsp; Someone who better understands why we're reaching that point without a multistep having been played will have to determine if this is the correct fix.

**Comment by thomas-maeder** [comment]
3 observations:
1. If we start Popeye with -maxmem 0, no hash table is allocated and the hashing machinery is never invoked. But Popeye still segfaults while solving the above examples; so fixing things with hashing disabled seems to be the right approach, as you (@JoshuaGreen) suggest in your more recent comments. We can hope that once Popeye sans hashing behaves correctly, we can re-enable hashing without further issues.
2. During a capture series, the en passant status is updated after each element of the series; updating it at the end of the series seems sufficient, though since a piece can't capture itself en passant, for different reasons.
3. The main difference introduced by CaptureSeries is that a single move's movements are spread over multiple plies. Some assumptions that were valid pre-CaptureSeries are no longer valid now.

**Comment by thomas-maeder** [comment]
I have dug a bit deeper: the reason for the integer underflow on line 190 is that we decrease en_passant_top for a different value of nbply than what we increased it for earlier. This is a consequence of the ply acrobatics that we are doing in the SeriesCapture implementation.

As a quick fix, we can remember the ply that we increase en_passant_top for, and use that value rather than the current nbply for decreasing.

A longer term fix will probably consist in adding a move effects undoer slice just before the slice that starts a new ply for the next element of the capture series.

**Comment by JoshuaGreen** [comment]
@thomas-maeder, good work on tracking that down, and good timing as I was about to submit a PR for my
```c
if (en_passant_was_multistep_played(nbply))
  --en_passant_top[nbply];
```
suggestion.&nbsp; As it is, I can now withdraw that suggestion and prefer
```c
assert(en_passant_was_multistep_played(nbply));
--en_passant_top[nbply];
```
to catch such issues going forward.

Considering either of your suggested fixes, either is beyond what logic I understand, but I do know that we have to keep in mind that the current en passant logic makes it difficult to adjust things for anything other than the final `ply`.&nbsp; The issue is that we store the capture squares as a series of adjacent ranges, and removing an element from the end of one range implicitly _adds_ it to the beginning of the following range.&nbsp; I suppose the simplest fix for _this_ would be to replace the "added" square with a dummy value that can be ignored during iteration, though that would likely complicate the iteration logic (if we haven't already prepared for this).&nbsp; Alternatively, we could redesign the en passant interface, say by instead storing the squares in a two-dimensional array with the first dimension of size `maxply` and the second dimension of size `nr_rows_on_board - 2` (or whatever the right value is, or maybe 1 more than that to allow a dummy value to always be used to end iteration).

**Comment by thomas-maeder** [comment]
Well, I'm not quite sure yet.

First of all, my quick fix seems to prevent most of the segfaults, but I still have a position with seriescapture and protean pieces where the
assert(en_passant_top[entry->u.ep_capture_potential.ply]>0);
that I have added in move_effect_journal_undo_remember_ep() fails.

And worse, it also fails for some of the example problems with total invisible pieces.

So there's more work to be done before we can decide what the real fix looks like.

**Comment by thomas-maeder** [comment]
I had forgotten about make sure that also move_effect_journal_redo_remember_ep()...

With this, things seem stable.

**Comment by JoshuaGreen** [comment]
@thomas-maeder, I'd like to better understand your [`433-segmentation-fault-in-seriescapture-%2B-proteanchess`](https://github.com/thomas-maeder/popeye/tree/433-segmentation-fault-in-seriescapture-%2B-proteanchess) branch.&nbsp; In particular, I'd like to understand how it avoids the issue I noted above.&nbsp; Specifically, suppose we have searched to a depth of 3 ply.&nbsp; We could presumably have `en_passant_top[0] == 0`, `en_passant_top[1] == 2`, `en_passant_top[2] == 4`, and `en_passant_top[3] == 6`.&nbsp; At this point, we can only safely manipulate `en_passant_top[3]` up and down; anything else will screw up the following range.&nbsp; Are these changes guaranteed to enforce this?

**Comment by thomas-maeder** [comment]
> I had always assumed that nbply was the maximum

That's what I meant with "ply acrobatics" above.

Your assumption is mostly right. In some situations, we use auxiliary plies because we need to generate and possibly play moves for some test (e.g. is there a cage in Cage Circe).

So the real maximum is ply_watermark (static in ply.c), which normally, but not quite always, is equal to nbply.

I SeriesCapture, we start a new ply for each movement in the series (except the initial capture which is done in the regular ply).

This is the first time that a single move can spread over multiple plies. Several places are not prepared for this situation, so we have to take some measures. STSeriesCaptureJournalFixer is one of them; it fakes that the entire capture series belongs to one ply - otherwise line oriented output (where we undo and then redo the effects of the current move, assuming there in the current ply) wouldn't work correctly IIRC.

Storing the ply with the en passant capture is now another such measure.

**Comment by JoshuaGreen** [comment]
Ugh.&nbsp; It certainly doesn't seem that the en passant infrastructure was designed for such shenanigans.&nbsp; Perhaps it should be reimagined.&nbsp; (Yes, that's my answer to _everything_.)&nbsp; I'm thinking that we need a new layer here:
moves are composed of `ply`s, but `ply`s are composed of ... what?&nbsp; Something like "piece_move"s ?&nbsp; Typically a `ply` would consist of exactly one "piece_move," but for SeriesCapture we need more of them.

(BTW, I'm having a hard time finding a definition for SeriesCapture.&nbsp; Does it mean that one must always begin with a capture and one is allowed to keep capturing, only _needing_ to stop when a move isn't a capture or when it delivers check?)

**Comment by thomas-maeder** [comment]
The award of this years Sake Tourney contains the definition:
https://problem-paradise.com/2023/09/13/the-award-of-the-20th-japanese-sake-tourney/

Everything you assume at the end of your comment is correct. In addition, every part of the series has to be a legal move; the move X*a2*e4*g6-h2 is not allowed if, say, X*a2*e4 is illegal.

**Comment by JoshuaGreen** [comment]
Based on my (limited) understanding of what's going on here, I believe we can tolerate stronger `assert`s.&nbsp; I've therefore modified your [`433-segmentation-fault-in-seriescapture-+-proteanchess`](https://github.com/thomas-maeder/popeye/tree/433-segmentation-fault-in-seriescapture-%2B-proteanchess) branch into [`433-segmentation-fault-in-seriescapture-+-proteanchess_extra-asserts`](https://github.com/thomas-maeder/popeye/tree/433-segmentation-fault-in-seriescapture-%2B-proteanchess_extra-asserts).&nbsp; Please let me know what you think.

**Comment by thomas-maeder** [comment]
@JoshuaGreen the assertions added in [433-segmentation-fault-in-seriescapture-+-proteanchess_extra-asserts] fail for the following example files:
- total_invisible.ref
- total_invisible.tst

So the assert()s or the implementation of total invisibles is/are wrong; that's an inclusive or, not an either-or - hence "is/are" :-)

py: pieces/walks/pawns/en_passant.c:219: Assertion `entry->u.ep_capture_potential.ply>=nbply' failed.

That's in ...redo... - if I change >= on line 219 to <=, I get:
py: pieces/walks/pawns/en_passant.c:198: Assertion `entry->u.ep_capture_potential.ply>=nbply' failed.

That's in ...undo....

**Comment by JoshuaGreen** [comment]
Fascinating.  I added that `assert`ion based on the following set of arguments:
1. `nbply` is a `ply` that is currently being tracked.
2. We can only update the en passant range corresponding to the last `ply` being tracked (as otherwise we'll screw up the following range).
3. We're trying to update the range corresponding to `ply` `entry->u.ep_capture_potential.ply`.
4. Therefore, `entry->u.ep_capture_potential.ply` should be &ge; `nbply`.

That `assert`ion is intended to capture 2, and I'm not exactly sure which of these could be wrong.&nbsp; Perhaps it's 1, in which case that `assert`ion should be replaced with something more appropriate (or dropped, if there's no obvious alternative).&nbsp; Alternatively, maybe 2 is wrong and there's some complex logic somewhere that allows us to adjust earlier `ply`s when needed; in that case the `assert`ion should simply be dropped.

**Comment by thomas-maeder** [comment]
I have now managed to move remembering en passant potential to a later point, when the entire capture series and the journal fixing are over.

This has allowed me to remove the code where we remember the ply with the en passant square; I decided to keep it only for potential assert()s.

I had to temporarily comment out these assert()s because the implementation of total invisibles fails them. I am convinced that this is a bug in the TI machinery that has always been there, but was only revealed by these assert()s.

I don't think that I have ever thought about en passant captures in the presence of total invisible pieces. Unless we find examples, we might consider forgetting en passant while figuring out moves by invisible pieces. But that's something for a different issue.

**Comment by JoshuaGreen** [comment]
I thought of plausible way that my point 2 could be safely violated.&nbsp; If we remove an en passant capture from an earlier range but then replace it with a new value then the alignment of the following range would be restored.&nbsp; I don't know if any of infrastructure relies on this type of paired operations, but if so we should probably replace them with a single operation.

**Comment by thomas-maeder** [comment]
Fixed for 4.91
